intake_bluesky/in_memory.py

```diff
@@ -48,22 +48,19 @@ def __init__(self, handler_registry=None, query=None, **kwargs):
         self._uid_to_run_start_doc = {}
         super().__init__(**kwargs)
 
-    def upsert(self, gen_func, gen_args, gen_kwargs):
-        gen = gen_func(*gen_args, **gen_kwargs)
-        name, run_start_doc = next(gen)
+    def upsert(self, gen_func, start_doc, stop_doc, gen_args, gen_kwargs):
 
-        if name != 'start':
-            raise ValueError("Expected a generator of (name, doc) pairs where "
-                             "the first entry was ('start', {...}).")
+        if not isinstance(start_doc, dict):
```

Comment:
Anything that duck-types as a `dict` should be acceptable here. This check could be done as `isinstance(start_doc, collections.abc.MutableMapping)`, which would return `True` for anything that implemented the _methods_ of `dict`.

But maybe it's worth pausing to consider if we want this kind of validation at all. What failure mode is it protecting us from?